File: packages/cli/src/commands/restore.ts
import { resolve } from "node:path";
import { readdir } from "node:fs/promises";
import { log } from "../lib/output";
import { loadRegistry } from "../lib/registry";
import { isContainerRunning } from "../lib/docker";
import type { ServiceName } from "../lib/types";
import * as postgres from "../services/postgres";
import * as redis from "../services/redis";
import * as rabbitmq from "../services/rabbitmq";

const BACKUPS_DIR = resolve(process.cwd(), "backups");

const SERVICE_CONTAINERS: Record<ServiceName, string> = {
  postgres: "infra-postgres",
  redis: "infra-redis",
  rabbitmq: "infra-rabbitmq",
};

function detectService(fileName: string): ServiceName | null {
  if (fileName.startsWith("postgres_")) return "postgres";
  if (fileName.startsWith("redis_")) return "redis";
  if (fileName.startsWith("rabbitmq_")) return "rabbitmq";
  return null;
}

async function findLatestBackup(
  appDir: string,
  service: ServiceName,
): Promise<string | null> {
  try {
    const files = await readdir(appDir);
    const matching = files
      .filter((f) => f.startsWith(`${service}_`))
      .sort()
      .reverse();
    return matching[0] ? resolve(appDir, matching[0]) : null;
  } catch {
    return null;
  }
}

async function restoreService(
  service: ServiceName,
  appName: string,
  filePath: string,
): Promise<void> {
  switch (service) {
    case "postgres":
      await postgres.restore(appName, filePath);
      break;
    case "redis":
      await redis.restore(appName, filePath);
      break;
    case "rabbitmq":
      await rabbitmq.restore(appName, filePath);
      break;
  }
}

export async function restoreCommand(
  appName: string,
  services: ServiceName[] | undefined,
  filePath: string | undefined,
  force: boolean,
): Promise<void> {
  if (!appName) {
    log.error("App name is required.");
    process.exit(1);
  }

  const registry = await loadRegistry();
  const app = registry.apps[appName];

  if (!app) {
    log.error(`App "${appName}" not found.`);
    process.exit(1);
  }

  // Restore from a specific file
  if (filePath) {
    const resolvedPath = resolve(filePath);
    const file = Bun.file(resolvedPath);
    if (!(await file.exists())) {
      log.error(`File not found: ${resolvedPath}`);
      process.exit(1);
    }

    const fileName = resolvedPath.split("/").pop() ?? "";
    const service = detectService(fileName);
    if (!service) {
      log.error(`Cannot detect service from filename "${fileName}". Expected format: postgres_*, redis_*, rabbitmq_*`);
      process.exit(1);
    }

    if (!(await isContainerRunning(SERVICE_CONTAINERS[service]))) {
      log.error(`Container "${SERVICE_CONTAINERS[service]}" is not running.`);
      process.exit(1);
    }

    if (!force) {
      const readline = await import("node:readline");
      const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
      const answer = await new Promise<string>((r) =>
        rl.question(`Restore ${service} for "${appName}" from ${fileName}? [y/N] `, r),
      );
      rl.close();
      if (answer.toLowerCase() !== "y") {
        log.info("Cancelled.");
        return;
      }
    }

    await restoreService(service, appName, resolvedPath);
    log.success(`${service} restored from ${fileName}`);
    return;
  }

  // Restore latest backups for each service
  const provisionedServices = Object.keys(app.services) as ServiceName[];
  const targetServices = services?.length
    ? services.filter((s) => provisionedServices.includes(s))
    : provisionedServices;

  const appDir = resolve(BACKUPS_DIR, appName);
  const restorePlan: { service: ServiceName; file: string }[] = [];

  for (const service of targetServices) {
    const latest = await findLatestBackup(appDir, service);
    if (latest) {
      restorePlan.push({ service, file: latest });
    } else {
      log.warn(`No backup found for ${service}`);
    }
  }

  if (restorePlan.length === 0) {
    log.error("No backups found to restore.");
    return;
  }

  if (!force) {
    log.title(`\nRestore plan for "${appName}":`);
    for (const { service, file } of restorePlan) {
      console.log(`  ${service} <- ${file.split("/").pop()}`);
    }
    console.log("");

    const readline = await import("node:readline");
    const rl = readline.createInterface({ input: process.stdin, output: process.stdout });
    const answer = await new Promise<string>((r) =>
      rl.question("Proceed with restore? [y/N] ", r),
    );
    rl.close();
    if (answer.toLowerCase() !== "y") {
      log.info("Cancelled.");
      return;
    }
  }

  for (const { service, file } of restorePlan) {
    if (!(await isContainerRunning(SERVICE_CONTAINERS[service]))) {
      log.error(`Container "${SERVICE_CONTAINERS[service]}" is not running. Skipping.`);
      continue;
    }

    try {
      await restoreService(service, appName, file);
      log.success(`${service} restored`);
    } catch (err) {
      log.error(`Failed to restore ${service}: ${err instanceof Error ? err.message : err}`);
    }
  }
}

File: internal/shelfcore/restore.go
package shelfcore

import (
	"context"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/ivan/infra-shelf/internal/docker"
	"github.com/ivan/infra-shelf/internal/registry"
	"github.com/ivan/infra-shelf/internal/services/aistor"
	"github.com/ivan/infra-shelf/internal/services/mongodb"
	"github.com/ivan/infra-shelf/internal/services/postgres"
	"github.com/ivan/infra-shelf/internal/services/rabbitmq"
	"github.com/ivan/infra-shelf/internal/services/redis"
)

type RestoreOptions struct {
	Services []string // empty = all backupable services on the app
}

// RestorePlanEntry describes one file the engine will restore from. Useful for
// dry-run confirmations in the CLI and for response payloads in the web UI.
type RestorePlanEntry struct {
	Service string
	File    string
}

// PlanRestore computes which backup files would be applied if RestoreApp were
// called with the same arguments. It does not touch any container.
func (e *Engine) PlanRestore(_ context.Context, appName string, opts RestoreOptions) ([]RestorePlanEntry, error) {
	if err := registry.ValidateAppName(appName); err != nil {
		return nil, err
	}

	reg, err := e.Store.Load()
	if err != nil {
		return nil, err
	}
	entry, ok := reg.Apps[appName]
	if !ok {
		return nil, fmt.Errorf("%w: %s", ErrAppNotFound, appName)
	}

	targets := chooseBackupServices(entry, opts.Services)
	appDir := filepath.Join(e.BackupsDir, appName)

	plan := []RestorePlanEntry{}
	for _, svc := range targets {
		latest := findLatestBackup(appDir, svc)
		if latest == "" {
			e.Reporter.Warn(fmt.Sprintf("no backup found for %s", svc))
			continue
		}
		plan = append(plan, RestorePlanEntry{Service: svc, File: latest})
	}
	if len(plan) == 0 {
		return nil, ErrBackupNotFound
	}
	return plan, nil
}

// RestoreApp restores the latest backup of each service in opts.Services
// (defaulting to every backupable service the app has provisioned). Use
// RestoreFromFile to restore a specific file.
func (e *Engine) RestoreApp(ctx context.Context, appName string, opts RestoreOptions) error {
	plan, err := e.PlanRestore(ctx, appName, opts)
	if err != nil {
		return err
	}

	for _, p := range plan {
		container := serviceContainer[p.Service]
		if !docker.IsContainerRunning(ctx, container) {
			e.Reporter.Error(fmt.Sprintf("container %q is not running. Skipping %s", container, p.Service))
			continue
		}
		if err := runRestore(ctx, p.Service, appName, p.File); err != nil {
			e.Reporter.Error(fmt.Sprintf("failed to restore %s: %v", p.Service, err))
			continue
		}
		e.Reporter.Success(fmt.Sprintf("%s restored", p.Service))
	}
	return nil
}

// RestoreFromFile restores a single backup file. The service is detected from
// the file name prefix (postgres_, redis_, rabbitmq_, aistor_).
func (e *Engine) RestoreFromFile(ctx context.Context, appName, filePath string) error {
	if err := registry.ValidateAppName(appName); err != nil {
		return err
	}

	reg, err := e.Store.Load()
	if err != nil {
		return err
	}
	if _, ok := reg.Apps[appName]; !ok {
		return fmt.Errorf("%w: %s", ErrAppNotFound, appName)
	}

	resolved, err := filepath.Abs(filePath)
	if err != nil {
		return err
	}
	if _, err := os.Stat(resolved); err != nil {
		return fmt.Errorf("file not found: %s", resolved)
	}

	fileName := filepath.Base(resolved)
	svc := detectService(fileName)
	if svc == "" {
		return fmt.Errorf("cannot detect service from filename %q (expected postgres_*, redis_*, rabbitmq_*, aistor_*, mongodb_*)", fileName)
	}

	if !docker.IsContainerRunning(ctx, serviceContainer[svc]) {
		return fmt.Errorf("%w: container %q for %s", ErrContainerNotRunning, serviceContainer[svc], svc)
	}

	if err := runRestore(ctx, svc, appName, resolved); err != nil {
		return err
	}
	e.Reporter.Success(fmt.Sprintf("%s restored from %s", svc, fileName))
	return nil
}

func runRestore(ctx context.Context, svc, appName, file string) error {
	switch svc {
	case "postgres":
		return postgres.Restore(ctx, appName, file)
	case "redis":
		return redis.Restore(ctx, appName, file)
	case "rabbitmq":
		return rabbitmq.Restore(ctx, appName, file)
	case "aistor":
		return aistor.Restore(ctx, appName, file)
	case "mongodb":
		return mongodb.Restore(ctx, appName, file)
	}
	return fmt.Errorf("unknown service: %s", svc)
}

func detectService(fileName string) string {
	for _, svc := range []string{"postgres", "redis", "rabbitmq", "aistor", "mongodb"} {
		if strings.HasPrefix(fileName, svc+"_") {
			return svc
		}
	}
	return ""
}

func findLatestBackup(appDir, svc string) string {
	entries, err := os.ReadDir(appDir)
	if err != nil {
		return ""
	}
	matching := []string{}
	prefix := svc + "_"
	for _, ent := range entries {
		if !ent.IsDir() && strings.HasPrefix(ent.Name(), prefix) {
			matching = append(matching, ent.Name())
		}
	}
	if len(matching) == 0 {
		return ""
	}
	sort.Sort(sort.Reverse(sort.StringSlice(matching)))
	return filepath.Join(appDir, matching[0])
}
